core-plugins/src/test/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSourceTest.java

```diff
@@ -0,0 +1,421 @@
+    try {
+      String xmlSourceFolder = "src/test/resources/";
+      String catalogLargeFile = "catalogLarge.xml";
+      String catalogSmallFile = "catalogSmall.xml";
+      FileUtils.copyFile(new File(xmlSourceFolder + catalogLargeFile), new File(SOURCE_FOLDER_PATH + catalogLargeFile));
+      FileUtils.copyFile(new File(xmlSourceFolder + catalogSmallFile), new File(SOURCE_FOLDER_PATH + catalogSmallFile));
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Method to clear source and target folders. This ensures that source and target folder are ready to use for next
+   * JUnit Test case.
+   */
+  private void clearSourceAndTargetFolder() {
+    File sourceDirectory = sourceFolder;
+    for (File sourceFile : sourceDirectory.listFiles()) {
+      sourceFile.delete();
+    }
+
+    File targetDirectory = targetFolder;
+    for (File targetFile : targetDirectory.listFiles()) {
+      targetFile.delete();
+    }
+  }
+
+  /**
+   * Method to Pre-Populate File tracking KeyValue with previously processed XML file.
+   * @param processedFileTable
+   * @param preProcessedDate
+   * @return
+   * @throws Exception
+   */
+  private Date createPreProcessedRecord(String processedFileTable, Date preProcessedDate) throws Exception {
+    DataSetManager<KeyValueTable> dataSetManager1 = getDataset(processedFileTable);
+    KeyValueTable keyValueTable = dataSetManager1.get();
+    //put one record of processed file.
+    File catalogLarge = new File(SOURCE_FOLDER_PATH + "catalogLarge.xml");
+    keyValueTable.write(catalogLarge.toURI().toString().getBytes(Charsets.UTF_8),
+                        String.valueOf(preProcessedDate.getTime()).getBytes(Charsets.UTF_8));
+    dataSetManager1.flush();
+    return preProcessedDate;
+  }
+
+  /**
+   * Method to return currently processed XML file list.
+   * @param processedFileTable
+   * @param preProcessedDate
+   * @return
+   * @throws Exception
+   */
+  private List<String> getProcessedFileList(String processedFileTable, Date preProcessedDate) throws Exception {
+    DataSetManager<KeyValueTable> dataSetManager = getDataset(processedFileTable);
+    KeyValueTable table = dataSetManager.get();
+    CloseableIterator<KeyValue<byte[], byte[]>> iterator = table.scan(null, null);
+
+    List<String> processedFileList  = new ArrayList<String>();
+
+    while (iterator.hasNext()) {
+      KeyValue<byte[], byte[]> keyValue = iterator.next();
+      Date date = new Date(Long.valueOf(new String(keyValue.getValue(), Charsets.UTF_8)));
+      if (date.after(preProcessedDate)) {
+        processedFileList.add(new String(keyValue.getKey(), Charsets.UTF_8));
+      }
+    }
+    return processedFileList;
+  }
+
+  @Test
+  /**
+   * This test validate following
+   * 1. Read multiple files from folder
+   * 2. Delete files once processed
+   * 3. Filter Pre-Processed file
+   */
+  public void testXMLReaderWithNoXMLPreProcessingRequired() throws Exception {
```

Comment:
can some of the common code across these tests be re-used by moving to helper methods instead of copying it over again?